File: cpp/src/observatory/operation/MessagingPingPongOperation.cpp
#include "MessagingPingPongOperation.h"

namespace Observatory {

MessagingPingPongOperation::MessagingPingPongOperation(Benchmark *benchmark, Benchmark::Mode mode, uint32_t operationCount, uint32_t operationSize) :
        LatencyOperation(benchmark, mode, operationCount, operationSize) {}

const char *MessagingPingPongOperation::getClassName() const {
    return "MessagingPingPongOperation";
}

const char* MessagingPingPongOperation::getOutputFilename() const {
    return "Messaging PingPong";
}

bool MessagingPingPongOperation::needsFilledReceiveQueue() const {
    return getMode() != Benchmark::Mode::SEND;
}

Status MessagingPingPongOperation::warmUp(uint32_t operationCount) {
    if(getMode() == Benchmark::Mode::SEND) {
        for(uint32_t i = 0; i < operationCount; i++) {
            Status status = getBenchmark().performPingPongIterationServer();

            if(status != Status::OK) {
                return status;
            }
        }
    } else {
        for(uint32_t i = 0; i < operationCount; i++) {
            Status status = getBenchmark().performPingPongIterationServer();

            if(status != Status::OK) {
                return status;
            }
        }
    }

    return Status::OK;
}

Status MessagingPingPongOperation::execute() {
    if (getMode() == Benchmark::Mode::SEND) {
        auto startTime = std::chrono::high_resolution_clock::now();

        for (uint32_t i = 0; i < getMeasurement().getOperationCount(); i++) {
            getMeasurement().startSingleMeasurement();
            Status status = getBenchmark().performPingPongIterationServer();
            getMeasurement().stopSingleMeasurement();

            if (status != Status::OK) {
                return status;
            }
        }

        uint64_t time = std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::high_resolution_clock::now() - startTime).count();
        getMeasurement().finishMeasuring(time);
    } else {
        for (uint32_t i = 0; i < getMeasurement().getOperationCount(); i++) {
            Status status = getBenchmark().performPingPongIterationServer();

            if (status != Status::OK) {
                return status;
            }
        }
    }

    return Status::OK;
}

}

File: java/benchmark/src/main/java/de/hhu/bsinfo/observatory/benchmark/MessagingPingPongOperation.java
package de.hhu.bsinfo.observatory.benchmark;

import de.hhu.bsinfo.observatory.benchmark.Benchmark.Mode;
import de.hhu.bsinfo.observatory.benchmark.result.Status;

public class MessagingPingPongOperation extends LatencyOperation {

    MessagingPingPongOperation(Benchmark benchmark, Mode mode, int operationCount, int operationSize) {
        super(benchmark, mode, operationCount, operationSize);
    }

    @Override
    String getOutputFilename() {
        return "Messaging PingPong";
    }

    @Override
    boolean needsFilledReceiveQueue() {
        return true;
    }

    @Override
    Status warmUp(int operationCount) {
        if(getMode() == Mode.SEND) {
            for(int i = 0; i < operationCount; i++) {
                Status status = getBenchmark().performPingPongIterationServer();

                if(status != Status.OK) {
                    return status;
                }
            }
        } else {
            for(int i = 0; i < operationCount; i++) {
                Status status = getBenchmark().performPingPongIterationClient();

                if(status != Status.OK) {
                    return status;
                }
            }
        }

        return Status.OK;
    }

    @Override
    Status execute() {
        if(getMode() == Mode.SEND) {
            long startTime = System.nanoTime();

            for(int i = 0; i < getMeasurement().getOperationCount(); i++) {
                getMeasurement().startSingleMeasurement();
                Status status = getBenchmark().performPingPongIterationServer();
                getMeasurement().stopSingleMeasurement();

                if(status != Status.OK) {
                    return status;
                }
            }

            getMeasurement().setTotalData(getMeasurement().getTotalData() * 2);
            getMeasurement().finishMeasuring(System.nanoTime() - startTime);
        } else {
            for(int i = 0; i < getMeasurement().getOperationCount(); i++) {
                Status status = getBenchmark().performPingPongIterationClient();

                if(status != Status.OK) {
                    return status;
                }
            }
        }

        return Status.OK;
    }
}
